embassy/src/channel/pubsub.rs

```diff
@@ -0,0 +1,651 @@
+                let message = queue_item.0.clone();
+
+                if current_message_index == 0 && queue_item.1 == 0 {
+                    s.queue.pop_front();
+                    s.publisher_wakers.iter_mut().flatten().for_each(|w| w.wake());
+                }
+
+                Some(WaitResult::Message(message))
+            }
+        })
+    }
+
+    unsafe fn register_subscriber_waker(&self, subscriber_index: usize, waker: &Waker) {
+        self.inner.lock(|inner| {
+            let mut s = inner.borrow_mut();
+            s.subscriber_wakers
+                .get_unchecked_mut(subscriber_index)
+                .as_mut()
+                .unwrap_unchecked()
+                .register(waker);
+        })
+    }
+
+    unsafe fn register_publisher_waker(&self, publisher_index: usize, waker: &Waker) {
+        self.inner.lock(|inner| {
+            let mut s = inner.borrow_mut();
+            s.publisher_wakers
+                .get_unchecked_mut(publisher_index)
+                .as_mut()
+                .unwrap_unchecked()
+                .register(waker);
+        })
+    }
+
+    unsafe fn unregister_subscriber(&self, subscriber_index: usize, subscriber_next_message_id: u64) {
+        self.inner.lock(|inner| {
+            let mut s = inner.borrow_mut();
+
+            // Remove the subscriber from the wakers
+            *s.subscriber_wakers.get_unchecked_mut(subscriber_index) = None;
+
+            // All messages that haven't been read yet by this subscriber must have their counter decremented
+            let start_id = s.next_message_id - s.queue.len() as u64;
+            if subscriber_next_message_id >= start_id {
+                let current_message_index = (subscriber_next_message_id - start_id) as usize;
+                s.queue
+                    .iter_mut()
+                    .skip(current_message_index)
+                    .for_each(|(_, counter)| *counter -= 1);
+            }
+        })
+    }
+
+    unsafe fn unregister_publisher(&self, publisher_index: usize) {
+        self.inner.lock(|inner| {
+            let mut s = inner.borrow_mut();
+            // Remove the publisher from the wakers
+            *s.publisher_wakers.get_unchecked_mut(publisher_index) = None;
+        })
+    }
+}
+
+/// Internal state for the PubSub channel
+struct PubSubState<T: Clone, const CAP: usize, const SUBS: usize, const PUBS: usize> {
+    /// The queue contains the last messages that have been published and a countdown of how many subscribers are yet to read it
+    queue: Deque<(T, usize), CAP>,
+    /// Every message has an id.
+    /// Don't worry, we won't run out.
+    /// If a million messages were published every second, then the ID's would run out in about 584942 years.
+    next_message_id: u64,
+    /// Collection of wakers for Subscribers that are waiting.  
+    /// The [Subscriber::subscriber_index] field indexes into this array.
+    subscriber_wakers: [Option<WakerRegistration>; SUBS],
```

Comment:
`WakerRegistration` is already kinda a `Option<Waker>`, no need to option-wrap it again.

Reply:
I knew that. The outer option is to keep track of which slot is populated or free.
I could make an enum `Slot` with `Taken(Waker)` and `Free`, but that seems a bit overkill to me.

Reply:
Yes, well, I do expect people to use them long term indeed. But I can imagine shorter lived ones too.
So to make it ergonomic, I think it should be either this API or a `Split`-like api where you can get them all once at the same time.

Personally I think this API is the better of the two and it only really costs an extra option per waker, so like one byte? (Unless the compiler is able to combine them)

Reply:
WakerRegistration is 4 bytes (without `executor-agnostic`) and has no niches, so `Option<WakerRegistration>` will be 8 bytes due to alignment.

there's also a 3rd possibility: allow creating them dynamically but not "recycle" slots, once you've created all N you can't create more even if you drop the old ones.

Reply:
Hmmmm, that doesn't *feel* right.
But you're right, an extra 4 bytes is a lot.

I could remove the options if I could query the WakerRegistration if they have a waker registered.
That would probably be the best solution right?

Instead of tracking a fixed index for each Sub/Pub, they just track the index they're using at that moment.
Then when a spot needs to be found, it just takes the first one that has no waker registered yet.

Reply:
That's not a bad idea at all! :) 

Maybe have it into a "MultiWakerRegistration" where you can register up to N wakers? It could be useful in other scenarios. 

Also, If registration is "smart" in that it doesn't register an already-registered waker, you don't even need to track indices in pub/subs at all, you just register the waker every time and that's it.

Reply:
> Also, If registration is "smart" in that it doesn't register an already-registered waker, you don't even need to track indices in pub/subs at all, you just register the waker every time and that's it.

You'd still need to know which one to remove on the drop. Oh and you'd need to be able to unregister a waker.

> Maybe have it into a "MultiWakerRegistration" where you can register up to N wakers? It could be useful in other scenarios.

Sure, that'd simplify some things in the pubsub impl.

Reply:
If you don't unregister, it's possible the `MultiWakerRegistration` will get full, yes. In that case you can wake+clear all, then register the waker you were trying to register. If the other tasks are still interested, they'll register themselves again and the thing will settle down in 1 iteration because there can't be more than N interested tasks. 

This is exactly what the current WakerRegistration does, it wakes the old waker if there's one, then replaces it. WakerRegistration is kinda a MultiWakerRegistration of size 1.

This may seem inefficient, but it should happen very rarely, and it yields the best code size vs trying to actively "clean up" no-longer-interested wakers.

Reply:
Ah right of course!
This saves another usize from the pubs/subs

Reply:
Alright, I changed it. It costs 8 bytes extra in the channel, but it saves 4 bytes per pub/sub capacity and 4 bytes per pub/sub instance. Well worth it.